File: data_structures/trie/trie.go
package datastructures

// TrieNode represents a single node in the trie.
type TrieNode struct {
	children [26]*TrieNode
	isEnd    bool
}

// Trie is a prefix tree data structure for efficient string operations.
type Trie struct {
	root *TrieNode
}

// NewTrie creates and returns a new Trie with an initialized root node.
func NewTrie() *Trie {
	return &Trie{root: &TrieNode{}}
}

// Insert adds a word to the trie.
func (t *Trie) Insert(word string) {
	node := t.root
	for _, ch := range word {
		idx := ch - 'a'
		if node.children[idx] == nil {
			node.children[idx] = &TrieNode{}
		}
		node = node.children[idx]
	}
	node.isEnd = true
}

// Search returns true if the exact word exists in the trie.
func (t *Trie) Search(word string) bool {
	node := t.findNode(word)
	return node != nil && node.isEnd
}

// StartsWith returns true if any word in the trie starts with the given prefix.
func (t *Trie) StartsWith(prefix string) bool {
	return t.findNode(prefix) != nil
}

// Delete removes a word from the trie. Returns false if the word is not found.
func (t *Trie) Delete(word string) bool {
	found := false
	t.deleteHelper(t.root, word, 0, &found)
	return found
}

// findNode traverses the trie following the given string and returns the final node,
// or nil if the path does not exist.
func (t *Trie) findNode(s string) *TrieNode {
	node := t.root
	for _, ch := range s {
		idx := ch - 'a'
		if node.children[idx] == nil {
			return nil
		}
		node = node.children[idx]
	}
	return node
}

// deleteHelper recursively deletes a word from the trie.
// Returns true if the parent should remove the reference to the current node.
func (t *Trie) deleteHelper(node *TrieNode, word string, depth int, found *bool) bool {
	if node == nil {
		return false
	}

	// Reached the end of the word
	if depth == len(word) {
		if !node.isEnd {
			return false
		}
		*found = true
		node.isEnd = false
		return !t.hasChildren(node)
	}

	idx := word[depth] - 'a'
	shouldDeleteChild := t.deleteHelper(node.children[idx], word, depth+1, found)

	if shouldDeleteChild {
		node.children[idx] = nil
		// Remove current node if it has no other children and is not end of another word
		return !node.isEnd && !t.hasChildren(node)
	}

	return false
}

// hasChildren returns true if the node has any non-nil children.
func (t *Trie) hasChildren(node *TrieNode) bool {
	for _, child := range node.children {
		if child != nil {
			return true
		}
	}
	return false
}

File: data_structures/trie/python/trie.py
from __future__ import annotations


class TrieNode:
    def __init__(self):
        self.children: list[TrieNode | None] = [None] * 26
        self.is_end: bool = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            idx = ord(ch) - ord('a')
            if node.children[idx] is None:
                node.children[idx] = TrieNode()
            node = node.children[idx]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def delete(self, word: str) -> bool:
        found = [False]
        self._delete_helper(self.root, word, 0, found)
        return found[0]

    def _find_node(self, s: str) -> TrieNode | None:
        node = self.root
        for ch in s:
            idx = ord(ch) - ord('a')
            if node.children[idx] is None:
                return None
            node = node.children[idx]
        return node

    def _delete_helper(self, node: TrieNode | None, word: str, depth: int, found: list[bool]) -> bool:
        if node is None:
            return False
        if depth == len(word):
            if not node.is_end:
                return False
            found[0] = True
            node.is_end = False
            return not self._has_children(node)
        idx = ord(word[depth]) - ord('a')
        should_delete = self._delete_helper(node.children[idx], word, depth + 1, found)
        if should_delete:
            node.children[idx] = None
            return not node.is_end and not self._has_children(node)
        return False

    def _has_children(self, node: TrieNode) -> bool:
        return any(child is not None for child in node.children)
